diff --git a/store/rootmulti/store.go b/store/rootmulti/store.go
index 85d9b5b20..9c55fed2c 100644
--- a/store/rootmulti/store.go
+++ b/store/rootmulti/store.go
@@ -198,6 +198,7 @@ func (rs *Store) Commit() types.CommitID {
 
 	// Need to update atomically.
 	batch := rs.db.NewBatch()
+	defer batch.Close()
 	setCommitInfo(batch, version, commitInfo)
 	setLatestVersion(batch, version)
 	batch.Write()
